components/brave_news/browser/channels_controller.cc

```diff
@@ -123,11 +127,24 @@ std::vector<std::string> ChannelsController::GetChannelLocales(
 
 void ChannelsController::GetAllChannels(ChannelsCallback callback) {
   publishers_controller_->GetOrFetchPublishers(base::BindOnce(
-      [](ChannelsCallback callback, PrefService* prefs, Publishers publishers) {
+      [](ChannelsController* controller, ChannelsCallback callback,
+         PrefService* prefs, Publishers publishers) {
         auto result = GetChannelsFromPublishers(publishers, prefs);
+        // fix this. result is a map
+        controller->enabled_channel_count_ = 0;
+        for (const auto& it : result) {
+          if (!it.second->subscribed_locales.empty()) {
+            controller->enabled_channel_count_++;
+          }
+        }
+        if (controller->news_metrics_) {
+          controller->news_metrics_->RecordTotalSubscribedCount(
+              p3a::SubscribeType::kChannels,
+              controller->enabled_channel_count_);
+        }
         std::move(callback).Run(std::move(std::move(result)));
       },
-      std::move(callback), base::Unretained(prefs_)));
+      base::Unretained(this), std::move(callback), base::Unretained(prefs_)));
```

Comment:
@bridiver mentioned that `base::Unretained` isn't actually safe here because `base::OneShotController` fires via PostTask.

That said, at the moment, everything else here is using `base::Unretained`, so not on you to fix it :smile:

Reply:
every wrong Unretained is a potential crash, so we definitely shouldn't increase the amount of potential failures. I haven't looked in details at lifetimes of these objects, but it looks scary indeed. Also given that channels_controller dies before publishers_controller

Reply:
switched to `WeakPtr` thanks